File: WQChartsDemo/Objc/ViewController/AreaChartVC.m
// https://github.com/CoderWQYao/WQCharts-iOS
//
// AreaChartVC.m
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

#import "AreaChartVC.h"

@interface AreaChartVC ()

@property (nonatomic, strong) WQAreaChartView* chartView;
@property (nonatomic, strong) UIColor* currentColor;

@end

@implementation AreaChartVC

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
    if (self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil]) {
        self.currentColor = Color_Blue;
    }
    return self;
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
    [self layoutRectangleChartView];
}

- (UIView *)createChartView {
    _chartView = [[WQAreaChartView alloc] init];
    return _chartView;
}

- (void)configChartOptions {
    [super configChartOptions];
    
    __weak typeof(self) weakSelf = self;
    
    [self.optionsView addItem:RadioCell.new
     .setTitle(@"FixedBounds")
     .setOptions(@[@"OFF",@"ON"])
     .setSelection(1)
     .setOnSelectionChange(^(RadioCell* cell, NSInteger selection) {
        WQAreaChartView* chartView = weakSelf.chartView;
        if(selection != 0) {
            chartView.chart.fixedMinY = @0;
            chartView.chart.fixedMaxY = @99;
        } else {
            chartView.chart.fixedMinY = nil;
            chartView.chart.fixedMaxY = nil;
        }
        [chartView redraw];
    })];
    
    [self.optionsView addItem:RadioCell.new
     .setTitle(@"Fill")
     .setOptions(@[@"OFF",@"ON",@"Gradient"])
     .setSelection(2)
     .setOnSelectionChange(^(RadioCell* cell, NSInteger selection) {
        [weakSelf updateFillPaint];
    })];
    
    [self.optionsView addItem:RadioCell.new
     .setTitle(@"Stoke")
     .setOptions(@[@"OFF",@"ON",@"Dash"])
     .setSelection(1)
     .setOnSelectionChange(^(RadioCell* cell, NSInteger selection) {
        WQAreaChartView* chartView = weakSelf.chartView;
        [weakSelf setupStrokePaint:chartView.chart.paint.stroke color:Color_White type:selection];
        [chartView redraw];
    })];
    
}

- (void)updateChartExchangeXY {
    [super updateChartExchangeXY];
    self.chartView.padding = [self chartViewPaddingForSelection:[self radioCellSelectionForKey:@"Padding"]];
    [self updateFillPaint];
    [self updateItems];
    [self.view setNeedsLayout];
}

- (void)updateFillPaint {
    NSInteger selection = [self radioCellSelectionForKey:@"Fill"];
    WQAreaChartView* chartView = self.chartView;
    WQAreaChart* chart = chartView.chart;
    WQChartFillPaint* paint = chart.paint.fill;
    switch (selection) {
        case 1: {
            paint.color = self.currentColor;
            paint.shader = nil;
            break;
        }
        case 2: {
            paint.color = nil;
            UIColor* color = self.currentColor;
            paint.shader = [[WQChartLinearGradient alloc] initWithStart:[chart convertRelativePointFromViewPoint:CGPointMake(0.5, 0)] end:[chart convertRelativePointFromViewPoint:CGPointMake(0.5, 1)] colors:@[[color colorWithAlphaComponent:0.1],color]];
            break;
        }
        default: {
            paint.color = nil;
            paint.shader = nil;
            break;
        }

    }
    [chartView redraw];
}


#pragma mark - Items

- (void)configChartItemsOptions {
    [super configChartItemsOptions];
    
    __weak typeof(self) weakSelf = self;
    
    [self.optionsView addItem:RadioCell.new
     .setTitle(@"ItemsHeaderText")
     .setOptions(@[@"OFF",@"ON"])
     .setSelection(1)
     .setOnSelectionChange(^(RadioCell* cell, NSInteger selection) {
        [weakSelf updateItems];
    })];
    
    [self.optionsView addItem:RadioCell.new
     .setTitle(@"ItemsFooterText")
     .setOptions(@[@"OFF",@"ON"])
     .setSelection(0)
     .setOnSelectionChange(^(RadioCell* cell, NSInteger selection) {
        [weakSelf updateItems];
    })];
    
}

- (NSMutableArray<WQAreaChartItem*>*)items {
    NSArray<WQAreaChartItem*>* items = self.chartView.chart.items;
    if (items) {
        return [NSMutableArray arrayWithArray:items];
    } else {
        NSMutableArray<WQAreaChartItem*>* items = [NSMutableArray array];
        for (NSInteger i=0; i<5; i++) {
           WQAreaChartItem* item = [self createItemAtIndex:i];
           [items addObject:item];
        }
        self.chartView.chart.items = items;
        return items;
    }
}

- (NSString*)itemsOptionTitle {
    return @"Items";
}

- (WQAreaChartItem*)createItemAtIndex:(NSInteger)index {
    WQAreaChartItem* item = [[WQAreaChartItem alloc] initWithValue:CGPointMake(index, [NSNumber randomIntegerFrom:0 to:99])];
    
    WQChartText* headerText = [[WQChartText alloc] init];
    headerText.font = [UIFont systemFontOfSize:9];
    headerText.color = Color_White;
    item.headerText = headerText;
    
    WQChartText* footerText = [[WQChartText alloc] init];
    footerText.font = [UIFont systemFontOfSize:9];
    footerText.color = Color_White;
    item.footerText = footerText;
    
    [self updateItem:item];
    return item;
}

- (SliderCell*)createItemCellWithItem:(WQAreaChartItem*)item atIndex:(NSInteger)index {
    __weak typeof(self) weakSelf = self;
    return SliderCell.new
    .setSliderValue(0,99,item.value.y)
    .setDecimalCount(0)
    .setObject(item)
    .setOnValueChange(^(SliderCell* cell,float value) {
        WQAreaChartItem* item = (WQAreaChartItem*)cell.object;
        item.value = CGPointMake(item.value.x, value);
        [weakSelf updateItem:item];
        [weakSelf.chartView redraw];
    });
}


- (void)itemsDidChange:(NSMutableArray<WQAreaChartItem*>*)items {
    self.chartView.chart.items = items;
    [self.chartView redraw];
}

- (void)updateItem:(WQAreaChartItem*)item {
    BOOL exchangeXY = self.chartView.chart.exchangeXY;
    
    WQChartText* headerText = item.headerText;
    if (headerText) {
        headerText.hidden = [self radioCellSelectionForKey:@"ItemsHeaderText"] == 0;
        WQChartTextBlocks* headerTextBlocks = WQChartTextBlocks.new;
        headerTextBlocks.offsetByAngle = ^CGFloat(WQChartText * _Nonnull chartText, CGSize size, CGFloat angle) {
            if (exchangeXY) {
                return 10;
            } else {
                return 10;
            }
        };
        headerText.delegateUsingBlocks = headerTextBlocks;
    }
    
    WQChartText* footerText = item.footerText;
    if (footerText) {
        footerText.hidden = [self radioCellSelectionForKey:@"ItemsFooterText"] == 0;
        WQChartTextBlocks* footerTextBlocks = WQChartTextBlocks.new;
        footerTextBlocks.offsetByAngle = ^CGFloat(WQChartText * _Nonnull chartText, CGSize size, CGFloat angle) {
            if (exchangeXY) {
                return 10;
            } else {
                return 10;
            }
        };
        footerText.delegateUsingBlocks = footerTextBlocks;
    }
    
}

- (void)updateItems {
    for (WQAreaChartItem* item in self.chartView.chart.items) {
        [self updateItem:item];
    }
    [self.chartView redraw];
}

#pragma mark - ChartViewDrawDelegate


- (void)chartViewWillDraw:(WQChartView *)chartView inRect:(CGRect)rect context:(CGContextRef)context {
    [super chartViewWillDraw:chartView inRect:rect context:context];
    
    for (WQAreaChartItem* item in self.chartView.chart.items) {
        NSString* string = [NSString stringWithFormat:@"%ld", (NSInteger)round(item.value.y)];
        item.headerText.string = string;
        item.footerText.string = string;
    }
}


- (void)appendAnimationKeys:(NSMutableArray<NSString *> *)animationKeys {
    [super appendAnimationKeys:animationKeys];
    [animationKeys addObject:@"Color"];
    [animationKeys addObject:@"Values"];
}

- (void)prepareAnimationOfChartViewForKeys:(NSArray<NSString*>*)keys {
    [super prepareAnimationOfChartViewForKeys:keys];
    
    WQAreaChart* chart = self.chartView.chart;
    
    if ([keys containsObject:@"Color"]) {
        UIColor* toColor = [self.currentColor isEqual:Color_Blue] ? Color_Red : Color_Blue;
        WQChartFillPaint* paint = chart.paint.fill;
        if (paint.color) {
            paint.colorTween = [[WQChartUIColorTween alloc] initWithFrom:paint.color to:toColor];
        }
        if (paint.shader) {
            WQChartLinearGradient* shader = (WQChartLinearGradient*)paint.shader;
            shader.colorsTween = [[WQChartUIColorArrayTween alloc] initWithFrom:shader.colors to:@[[toColor colorWithAlphaComponent:0.1],toColor]];
        }
        self.currentColor = toColor;
    }
    
    if ([keys containsObject:@"Values"]) {
        [self.chartView.chart.items enumerateObjectsUsingBlock:^(WQAreaChartItem * _Nonnull item, NSUInteger idx, BOOL * _Nonnull stop) {
            item.valueTween = [[WQChartCGPointTween alloc] initWithFrom:item.value to:CGPointMake(item.value.x, [NSNumber randomIntegerFrom:0 to:99])];
        }];
    }
    
}

#pragma mark - AnimationDelegate

- (void)animation:(WQChartAnimation *)animation progressDidChange:(CGFloat)progress {
    [super animation:animation progressDidChange:progress];
    
    if (animation.animatable == self.chartView) {
        [self.chartView.chart.items enumerateObjectsUsingBlock:^(WQAreaChartItem * _Nonnull item, NSUInteger idx, BOOL * _Nonnull stop) {
            [self updateSliderValue:item.value.y forKey:@"Items" atIndex:idx];
        }];
    }
    
}

@end

File: WQChartsDemo/Swift/ViewController/AreaChartVC.swift
// https://github.com/CoderWQYao/WQCharts-iOS
//
// AreaChartVC.swift
// WQCharts
//
// Created by WQ.Yao on 2020/01/02.
// Copyright (c) 2020年 WQ.Yao All rights reserved.
//

import UIKit

class AreaChartVC: CoordinateChartVC<AreaChartView>, ItemsOptionsDelegate {
    
    var currentColor: UIColor = Color_Blue
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        layoutRectangleChartView()
    }
    
    override func configChartOptions() {
        super.configChartOptions()
        
        optionsView.addItem(RadioCell()
            .setTitle("FixedBounds")
            .setOptions(["OFF","ON"])
            .setSelection(1)
            .setOnSelectionChange({[weak self](cell, selection) in
                guard let self = self else {
                    return
                }
                let chartView = self.chartView
                if selection != 0 {
                    chartView.chart.fixedMinY = 0
                    chartView.chart.fixedMaxY = 99
                } else {
                    chartView.chart.fixedMinY = nil
                    chartView.chart.fixedMaxY = nil
                }
                chartView.redraw()
            })
        )
        
        optionsView.addItem(RadioCell()
            .setTitle("Fill")
            .setOptions(["OFF","ON","Gradient"])
            .setSelection(2)
            .setOnSelectionChange({[weak self](cell, selection) in
                guard let self = self else {
                    return
                }
                let chartView = self.chartView
                guard let paint = chartView.chart.paint?.fill else {
                    return
                }
                switch (selection) {
                case 1:
                    paint.color = self.currentColor
                    paint.shader = nil
                case 2:
                    paint.color = self.currentColor
                    paint.shader = {(paint, path, object) -> Shader? in
                        guard let color = paint.color else {
                            return nil
                        }
                        let graphic = object as! AreaGraphic
                        return LinearGradientShader(graphic.stringStart, graphic.stringEnd, [color.withAlphaComponent(0.1),color], [0,1])
                    }
                default:
                    paint.color = nil
                    paint.shader = nil
                }
                chartView.redraw()
            })
        )
        
        optionsView.addItem(RadioCell()
            .setTitle("Stoke")
            .setOptions(["OFF","ON","Dash"])
            .setSelection(1)
            .setOnSelectionChange({[weak self](cell, selection) in
                guard let self = self else {
                    return
                }
                let chartView = self.chartView
                
                self.setupStrokePaint(chartView.chart.paint?.stroke, Color_White, selection)
                chartView.redraw()
            })
        )
        
    }
    
    override func updateChartExchangeXY() {
        super.updateChartExchangeXY()
        chartView.padding = chartViewPadding(forSelection: radioCellSelectionForKey("Padding"))
        updateItems()
        view.setNeedsLayout()
    }
    
    // MARK: - Items
    
    override func configChartItemsOptions() {
        super.configChartItemsOptions()
        
        optionsView.addItem(RadioCell()
            .setTitle("ItemsHeaderText")
            .setOptions(["OFF","ON"]).setSelection(1)
            .setOnSelectionChange({[weak self](cell, selection) in
                self?.updateItems()
            })
        )
        
        optionsView.addItem(RadioCell()
            .setTitle("ItemsFooterText")
            .setOptions(["OFF","ON"])
            .setSelection(0)
            .setOnSelectionChange({[weak self](cell, selection) in
                self?.updateItems()
            })
        )
        
    }
    
    var items: NSMutableArray {
        if let items = chartView.chart.items {
            return NSMutableArray.init(array: items)
        } else {
            let items = NSMutableArray()
            for i in 0..<5 {
                let item = createItem(atIndex: i)!
                items.add(item)
            }
            chartView.chart.items = (items as! [AreaChartItem])
            return items
        }
    }
    
    var itemsOptionTitle: String {
        return "Items"
    }
    
    func createItem(atIndex index: Int) -> Any? {
        let item = AreaChartItem(CGPoint(x: CGFloat(index), y: CGFloat.random(in: 0...99)))
        
        let headerText = ChartText()
        headerText.font = UIFont.systemFont(ofSize: 9)
        headerText.color = Color_White
        item.headerText = headerText
        
        let footerText = ChartText()
        footerText.font = UIFont.systemFont(ofSize: 9)
        footerText.color = Color_White
        item.footerText = footerText
        
        updateItem(item)
        return item
    }
    
    func createItemCell(withItem item: Any, atIndex index: Int)  -> UIView {
        return SliderCell()
            .setObject(item)
            .setSliderValue(0,99,(item as! AreaChartItem).value.y)
            .setDecimalCount(0)
            .setOnValueChange {[weak self](cell, value) in
                guard let self = self, let item = cell.object as? AreaChartItem else {
                    return
                }
                
                item.value.y = value
                self.updateItem(item)
                self.chartView.redraw()
        }
    }
    
    func itemsDidChange(_ items: NSMutableArray) {
        chartView.chart.items = (items as! [AreaChartItem])
        chartView.redraw()
    }
    
    func updateItem(_ item: AreaChartItem) {
        let exchangeXY = chartView.chart.exchangeXY
        
        let string = String(format: "%ld", Int(round(item.value.y)))
        
        if let headerText = item.headerText {
            headerText.string = string
            headerText.hidden = radioCellSelectionForKey("ItemsHeaderText") == 0
            headerText.textOffsetByAngle = {(text, size, angle) -> CGFloat in
                if exchangeXY {
                    return 10
                } else {
                    return 10
                }
            }
        }
        
        if let footerText = item.footerText {
            footerText.string = string
            footerText.hidden = radioCellSelectionForKey("ItemsFooterText") == 0
            footerText.textOffsetByAngle = {(text, size, angle) -> CGFloat in
                if exchangeXY {
                    return 10
                } else {
                    return 10
                }
            }
        }
    }
    
    func updateItems() {
        let chartView = self.chartView
        if let items = chartView.chart.items {
            for item in items {
                updateItem(item)
            }
        }
        chartView.redraw()
    }
    
    // MARK: - ChartViewDrawDelegate
    
    override func chartViewWillDraw(_ chartView: ChartView, inRect rect: CGRect, context: CGContext) {
        super.chartViewWillDraw(chartView, inRect: rect, context: context)
        
        if let items = self.chartView.chart.items{
            for item in items {
                let string = String(format: "%ld", Int(round(item.value.y)))
                item.headerText?.string = string
                item.footerText?.string = string
            }
        }
        
    }
    
    // MARK: - Animation
    
    override func appendAnimationKeys(_ animationKeys: NSMutableArray) {
        super.appendAnimationKeys(animationKeys)
        animationKeys.add("Color")
        animationKeys.add("Values")
    }
    
    override func prepareAnimationOfChartView(forKeys keys: [String]) {
        super.prepareAnimationOfChartView(forKeys: keys)
        
        let chart = self.chartView.chart
        
        if keys.contains("Color"), let paint = chart.paint?.fill {
            let color = currentColor.isEqual(Color_Blue) ? Color_Red : Color_Blue
            paint.transformColor = TransformUIColor(paint.color ?? .clear, color)
            currentColor = color
            
            let cell = findRadioCellForKey("Fill")!
            if cell.selection == 0 {
                cell.selection = 1
            }
        }
        
        if keys.contains("Values") {
            if let items = chart.items {
                for item in items {
                    item.transformValue = TransformCGPoint(item.value, CGPoint(x: item.value.x, y: CGFloat(Int.random(in: 0...99))))
                }
            }
        }
        
    }
    
    // MARK: - AnimationDelegate
    
    override func animation(_ animation: Animation, progressDidChange progress: CGFloat) {
        super.animation(animation, progressDidChange: progress)
        
        if chartView.isEqual(animation.transformable) {
            if let items = chartView.chart.items {
                for (idx, item) in items.enumerated() {
                    updateSliderValue(item.value.y, forKey: "Items", atIndex: idx)
                }
            }
        }
        
    }
    
}
